dashboard/pages/visualise_time_changes.py

```diff
@@ -0,0 +1,116 @@
+"""Script to create visualisations of changes over time"""
+
+from datetime import datetime
+import streamlit as st
+import pandas as pd
+import altair as alt
+from db_functions import get_scores_topic, get_topic_names
+
+
+def resample_dataframe(df: pd.DataFrame, time_interval: str, avg_over_source: bool):
+    """Resamples the dataframe to return the average sentiment scores by source, topic 
+    over a set of grouped time intervals"""
+
+    df['date_published'] = pd.to_datetime(df['date_published'])
+
+    if avg_over_source:
+        df["source_name"] = "Average"
+    df_avg = df.groupby(['source_name', 'topic_name']).resample(
+        time_interval, on='date_published').mean().reset_index()
+
+    return df_avg
+
+
+def generate_warning_message(source_to_topics: dict) -> str:
+    """Generates a warning message that some sources don't cover some topics"""
+
+    if not source_to_topics:
+        return ""
+    return "\n".join([
+        f"""WARNING: {s} doesn't have any articles for {','.join(t)}"""
+        for s, t in source_to_topics])
+
+
+def visualise_change_over_time(df: pd.DataFrame, by_title: bool) -> alt.Chart:
+    """Visualise changes in sentiment over time"""
+
+    base = alt.Chart(df).encode(
+        alt.Color("source_name:N", title='Source Name').legend(None)
+    ).properties(
+        width=500
+    )
+
+    if not by_title:
+        y_axis = ('content_polarity_score', "Content Polarity Score")
+    else:
+        y_axis = ('title_polarity_score', "Title Polarity Score")
+
+    line = base.mark_line().encode(
+        x=alt.X('date_published:T', axis=alt.Axis(
+            offset=-150, title='Date Published', titleAnchor="end")),
+        y=alt.Y(f'{y_axis[0]}:Q', scale=alt.Scale(
+            domain=[-1, 1]), title=y_axis[1]),
+        tooltip=[alt.Tooltip(field="source_name", title="Source Name"),
+                 alt.Tooltip(field=f"{y_axis[0]}",
+                             title=f"Average {y_axis[1]}")]
+    ).properties(
+        width=500
+    )
```

Comment:
very small thing, but i believe you can add .interactive() just after the properties and it allows you to zoom in/move!

Reply:
good point - will experiment with doing that :)